### PR TITLE
KAFKA-15874: Add metric and request log attribute for deprecated request api versions (KIP-896)

### PR DESCRIPTION
**Pull request by ijuma**
Breakdown of this PR:
* Extend the generator to support deprecated api versions
* Set deprecated api versions via the request json files
* Expose the information via metrics and the request log

The relevant section of the KIP:

> * Introduce metric `kafka.network:type=RequestMetrics,name=DeprecatedRequestsPerSec,request=(api-name),version=(api-version),clientSoftwareName=(client-software-name),clientSoftwareVersion=(client-software-version)`
> * Add boolean field `requestApiVersionDeprecated`  to the request header section of the request log (alongside `requestApiKey` , `requestApiVersion`, `requestApiKeyName` , etc.).

Unit tests were added to verify the new generator functionality, the new metric and the new request log
attribute.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/network/RequestConvertToJsonTest.scala [hunk 1]
```diff
@@ -118,6 +119,21 @@ class RequestConvertToJsonTest {
     assertEquals(expectedNode, actualNode)
   }
 
+  @Test
+  def testRequestHeaderNodeWithDeprecatedApiVersion(): Unit = {
+    val alterIsrRequest = FetchRequest.Builder.forConsumer(0, 0, 0, Collections.emptyMap()).build(0);
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/network/RequestConvertToJsonTest.scala]: nit: `fetchRequest`?

File: generator/src/main/java/org/apache/kafka/message/StructRegistry.java [hunk 2]
```diff
@@ -110,6 +110,7 @@ private void addStructSpecs(Versions parentVersions, List<FieldSpec> fields) {
                     // Synthesize a StructSpec object out of the fields.
                     StructSpec spec = new StructSpec(typeName,
                         field.versions().toString(),
+                        null, // version deprecations not supported at field level //FIXME Perhaps we should pass `Versions.NONE.toString` instead
```
Comment (hachikuji) [inline generator/src/main/java/org/apache/kafka/message/StructRegistry.java]: Do you intend to address the FIXME in this patch?

**Review by hachikuji — COMMENTED** [review]
Looks good. Left a couple nits. We seem to be missing deprecations for `CreateAcls` and `DeleteAcls`.

File: core/src/test/scala/unit/kafka/network/RequestConvertToJsonTest.scala (hunk 1 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/network/RequestConvertToJsonTest.scala]: Fixed, thanks.

File: generator/src/main/java/org/apache/kafka/message/StructRegistry.java (hunk 2 above)
Comment (ijuma) [reply generator/src/main/java/org/apache/kafka/message/StructRegistry.java]: Yes! My bad for not fixing this before I asked for the review. I attempted a fix, let's see if any tests fail.

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM

**Comment by ijuma** [comment]
There are a few unrelated failures in the CI results (the set of tests that failed during the two runs is disjoint for the Java 21 build, for example).

**Comment by ijuma** [comment]
Merged to master and cherry-picked to 3.7.

**Pull request merged**